driver/src/main/java/org/neo4j/driver/internal/cluster/RoutingPooledConnection.java

```diff
@@ -0,0 +1,289 @@
+        {
+            throw handledException( e );
+        }
+    }
+
+    @Override
+    public void sync()
+    {
+        try
+        {
+            delegate.sync();
+        }
+        catch ( RuntimeException e )
+        {
+            throw handledException( e );
+        }
+    }
+
+    @Override
+    public void flush()
+    {
+        try
+        {
+            delegate.flush();
+        }
+        catch ( RuntimeException e )
+        {
+            throw handledException( e );
+        }
+    }
+
+    @Override
+    public void receiveOne()
+    {
+        try
+        {
+            delegate.receiveOne();
+        }
+        catch ( RuntimeException e )
+        {
+            throw handledException( e );
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        delegate.close();
+    }
+
+    @Override
+    public boolean isOpen()
+    {
+        return delegate.isOpen();
+    }
+
+    @Override
+    public void onError( Runnable runnable )
+    {
+        delegate.onError( runnable );
+    }
+
+    @Override
+    public boolean hasUnrecoverableErrors()
+    {
+        return delegate.hasUnrecoverableErrors();
+    }
+
+    @Override
+    public boolean isAckFailureMuted()
+    {
+        return delegate.isAckFailureMuted();
+    }
+
+    @Override
+    public ServerInfo server()
+    {
+        return delegate.server();
+    }
+
+    @Override
+    public BoltServerAddress boltServerAddress()
+    {
+        return delegate.boltServerAddress();
+    }
+
+    @Override
+    public Logger logger()
+    {
+        return delegate.logger();
+    }
+
+    @Override
+    public long lastUsedTimestamp()
+    {
+        return delegate.lastUsedTimestamp();
+    }
+
+    @Override
+    public void dispose()
+    {
+        delegate.dispose();
+    }
+
+    private RuntimeException handledException( RuntimeException e )
+    {
+        if ( e instanceof ServiceUnavailableException )
+        {
+            return handledServiceUnavailableException( ((ServiceUnavailableException) e) );
+        }
+        else if ( e instanceof ClientException )
+        {
+            return handledClientException( ((ClientException) e) );
+        }
+        else
+        {
+            return e;
+        }
+    }
+
+    private RuntimeException handledServiceUnavailableException( ServiceUnavailableException e )
+    {
+        BoltServerAddress address = boltServerAddress();
+        errorHandler.onConnectionFailure( address );
+        return new SessionExpiredException( format( "Server at %s is no longer available", address ), e );
+    }
+
+    private RuntimeException handledClientException( ClientException e )
+    {
+        if ( isFailureToWrite( e ) )
+        {
+            // The server is unaware of the session mode, so we have to implement this logic in the driver.
+            // In the future, we might be able to move this logic to the server.
+            switch ( accessMode )
+            {
+            case READ:
```

Comment:
It is possible to get `NotALeader` when trying to write in a read session with connection towards follower and `ForbiddenOnReadOnlyDatabase` when connection is towards read replica. So it looks like we still need this switch on the access mode.